handlers/handler-loadbalance/src/main/java/org/apache/servicecomb/loadbalance/filter/PriorityInstancePropertyDiscoveryFilter.java

```diff
@@ -0,0 +1,195 @@
+import org.apache.servicecomb.registry.discovery.DiscoveryTreeNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.netflix.config.DynamicPropertyFactory;
+
+/**
+ * Instance property with priority filter
+ */
+public class PriorityInstancePropertyDiscoveryFilter extends AbstractDiscoveryFilter {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(PriorityInstancePropertyDiscoveryFilter.class);
+
+  private static final String ALL_INSTANCE = "allInstance";
+
+  private String propertyKey;
+
+  @Override
+  protected void init(DiscoveryContext context, DiscoveryTreeNode parent) {
+    propertyKey = DynamicPropertyFactory.getInstance()
+        .getStringProperty("servicecomb.loadbalance.filter.priorityInstanceProperty.key", "environment")
+        .get();
+
+    // group all instance by property
+    Map<String, MicroserviceInstance> instances = parent.data();
+    Map<String, Map<String, MicroserviceInstance>> groupByProperty = new HashMap<>();
+    for (MicroserviceInstance microserviceInstance : instances.values()) {
+      String propertyValue = new PriorityInstanceProperty(propertyKey, microserviceInstance).getPropertyValue();
+      groupByProperty.computeIfAbsent(propertyValue, key -> new HashMap<>())
+          .put(microserviceInstance.getInstanceId(), microserviceInstance);
+    }
+    Map<String, DiscoveryTreeNode> children = new HashMap<>();
+    for (Map.Entry<String, Map<String, MicroserviceInstance>> entry : groupByProperty.entrySet()) {
+      children.put(entry.getKey(),
+          new DiscoveryTreeNode().subName(parent, entry.getKey()).data(entry.getValue()));
+    }
+    children.put(ALL_INSTANCE, new DiscoveryTreeNode().subName(parent, ALL_INSTANCE).data(instances));
+    parent.children(children);
+  }
+
+  @Override
+  protected String findChildName(DiscoveryContext context, DiscoveryTreeNode parent) {
+    Invocation invocation = context.getInputParameters();
+
+    // context property has precedence over instance property
+    String initPropertyValue = invocation.getContext()
+        .computeIfAbsent("x-" + propertyKey,
+            key -> new PriorityInstanceProperty(propertyKey, RegistrationManager.INSTANCE.getMicroserviceInstance())
+                .getPropertyValue());
+
+    PriorityInstanceProperty currentProperty = context.getContextParameter(propertyKey);
+    // start with initial value, then search with priority
+    if (Objects.isNull(currentProperty)) {
+      currentProperty = new PriorityInstanceProperty(propertyKey, initPropertyValue);
+      while (!parent.children().containsKey(currentProperty.getPropertyValue())
+          && currentProperty.hasChildren()) {
+        currentProperty = currentProperty.child();
+      }
+    } else {
+      if (currentProperty.hasChildren()) {
+        currentProperty = currentProperty.child();
+      }
+    }
+    LOGGER.info("Discovery instance filter by {}", currentProperty.toString());
```

Comment:
Maybe change to debug is better, or will print a log for each invocation.

Reply:
fixed